cupy/functional/piecewise.py

```diff
@@ -0,0 +1,84 @@
+import cupy
+
+from cupy import core
+
+_piecewise_krnl = core.ElementwiseKernel(
+    'U condlist, T funclist',
+    'T y',
+    ' if(condlist) y = funclist',
+    'piecewise_kernel'
+)
+
+
+def piecewise(x, condlist, funclist):
+    """Evaluate a piecewise-defined function.
+
+        Args:
+            x (cupy.ndarray): input domain
+            condlist (list of cupy.ndarray or bool scalars):
+                Each boolean array/ scalar corresponds to a function
+                in funclist. Length of funclist is equal to that of
+                condlist. If one extra function is given, it is used
+                as the default value when the otherwise condition is met
+            funclist (cupy.ndarray or list): list of scalar functions.
+
+        Returns:
+            cupy.ndarray: the scalar values in funclist on portions of x
+                defined by condlist.
+
+        .. warning::
+
+            This function currently doesn't support callable functions,
+            args and kw parameters.
+
+        .. seealso:: :func:`numpy.piecewise`
+        """
+    diffshape = 0
+    if cupy.isscalar(condlist):
+        condlist = [condlist]
+    elif isinstance(condlist[0], tuple) and x.ndim != 0:
+        diffshape = 2
+    elif cupy.isscalar(condlist[0]) and x.ndim != 0:
+        diffshape = 1
+    if not isinstance(condlist[0], cupy.ndarray):
+        condlist = cupy.asarray(condlist)
+    condlen = len(condlist)
+    if diffshape:
+        if x.ndim == condlist.ndim:
```

Comment:
Please check for `x.ndim != condlist.ndim` case.

Reply:
Isn't it already checked in this else ?
``` 
else:
    raise IndexError('boolean index did not match indexed array along '
    'dimension 0; dimension is {} but corresponding '
    'boolean dimension is {}'
    .format(x.shape[0], condlen))
```
I mean for example the first case will pass and the second will raise the error
```
x = testing.shaped_arange((2, 3), xp, dtype)
condlist = [False, True]
funclist = [1, 2]
xp.piecewise(x, condlist, funclist)
```
```
x = testing.shaped_arange((2, 3), xp, dtype)
condlist = [False, True, True]
funclist = [1, 2]
xp.piecewise(x, condlist, funclist)
```

Reply:
Could you explain why should we check the shape equality only when `x.ndim == condlist.ndim`?

Reply:
One more question. Why you don't check the shape when `diffshape == 0`?

Reply:
> Could you explain why should we check the shape equality only when `x.ndim == condlist.ndim`?

I thought I explained it by the above examples. When ``` x.ndim == condlist.ndim ```, we should check for the shapes. By not satisfying (x.dim == conlist.ndim) condition, only 2 options are possible now, either ``` condlen == x.shape[0] ``` (and in this case ``` condlen ``` should be adjusted to 1 to skip the error in L75 or simply to mean that a special broadcasting will be used for this case) or raise the error because of index mismatch along dimension 0 as it can be seen from the examples I explained [here](https://github.com/cupy/cupy/pull/3329#discussion_r436331763)
Does this answer the question ?

Reply:
Why do you check the shape equality only when `x.ndim == condlist.ndim`, and don't check when `x.ndim != condlist.ndim`?

Reply:
> One more question. Why you don't check the shape when `diffshape == 0`?

diffshape flag (as by its name) already targets special cases that need a different broadcasting ( list of scalars with multidimensional input or list of tuples/ nested tuples with multidimensional input). Handling these special cases require a pre-step before resuming the normal behavior for all (diffshape == 1, 2 or 0). This extra step can be seen by entering ```if diffshape: ``` condition. After that, the default behaviour is applied for different inputs regardless of diffshape value ( that can be seen from line 62, and of course with the exception of transposing condlist in case of diffshape == 1 later on...)

Reply:
> Why do you check the shape equality only when `x.ndim == condlist.ndim`, and don't check when `x.ndim != condlist.ndim`?

because I don't need the shapes information when x.ndim !=  condlist.ndim
Either condlen == x.shape[0] and here special broadcasting is used 
``` 
x = testing.shaped_arange((2, 3), xp, dtype)
condlist = [False, True]
funclist = [1, 2]
xp.piecewise(x, condlist, funclist) 
```
or raise error because of mismatch along dimension 0
``` x = testing.shaped_arange((2, 3), xp, dtype)
condlist = [False, True, True]
funclist = [1, 2]
xp.piecewise(x, condlist, funclist)
```
I only care about the shape along dimension 0 when x.ndim !=  condlist.ndim

Reply:
I found some different behaviors between this implementation and NumPy's one.
- Only cupy raises `ValueError`.
```py
    @testing.for_all_dtypes()
    @testing.numpy_cupy_array_equal()
    def test_broadcast(self, xp, dtype):
        x = testing.shaped_random(shape=(2, 3, 5), xp=xp, dtype=dtype)
        condlist = [[[True, False, False], [True, True, True]]]
        funclist = [-1, 2]
        return xp.piecewise(x, condlist, funclist)
```
- Only numpy raises error.
```py
    @testing.for_all_dtypes()
    @testing.numpy_cupy_array_equal()
    def test_broadcast(self, xp, dtype):
        x = testing.shaped_random(shape=(2, 3, 5), xp=xp, dtype=dtype)
        condlist = [[[True, False, False, False, False]], [[True, True, True, True, True]]]
        funclist = [-1, 2]
        return xp.piecewise(x, condlist, funclist)
```

Reply:
I will look into it